### PR TITLE
change python requirements to 3.10 - 3.12

### PR DESCRIPTION
**Pull request by EdwinB12**

### CODE REVIEW

**Comment by EdwinB12** [comment]
Checklist: 

- [ ] Put on minimum versions of python dependancies :  resolve #119
- [ ] Remove unneccessary python installs in dockerfiles 
- [ ] Change python versions in pyproject.toml and readme to document supported python versions : #124  
- [ ] Change github actions to run tests on all python versions
- [ ] Remove easy warnings : resolves #125 
- [ ] Consider building backend and frontend apps with all supported python versions

**Comment by EdwinB12** [comment]
@mhauru @GiorgioCerro - Could you look at the warnings spat out from the tests and see if there are any super easy ones you can address? 

You may need to flag the warnings as errors to see the full warning: 
`pytest -W error tests`

I'm going to merge this branch with test-actions branch to run the docker containers and check whether those actions pass okay.

**Comment by EdwinB12** [comment]
All actions have passed (including docker deployments on test-actions branch). 

I'm essentially happy for this to be merged after review as lots of the warnings have now been supressed. However, please feel free to have a go at some of the remaining warnings. 

Last thing @mhauru - We discussed whether we want to run a docker deployment for each python version. I think we leave this for now. I'm quite happy that the tests are being run for all supported python versions but let me know what you think.

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -4,12 +4,15 @@ on:
   push:
     branches: [main, develop, test-actions]
   pull_request:
-    branches: [main, develop, test-actions]
```
Comment (mhauru) [inline .github/workflows/tests.yaml:7]: Any particular reason to remove this here but not in the docker actions?

File: pyproject.toml [hunk 2]
```diff
@@ -89,19 +89,16 @@ target-version = "py310"
 
 [tool.ruff.lint]
 # TODO Comment in more of the below as we clean up the code.
-ignore = [
```
Comment (mhauru) [inline pyproject.toml:92]: This seems like its reverting a change from an earlier PR, which I assume is not intended.

**Review by mhauru — CHANGES_REQUESTED** [review]
Mostly looks great, I added one question and one request for change.

**Comment by mhauru** [comment]
>Last thing @mhauru - We discussed whether we want to run a docker deployment for each python version. I think we leave this for now. I'm quite happy that the tests are being run for all supported python versions but let me know what you think.

Sounds good.

The SQLAlchemy warning came from a legitimate bug (nice warning SQLAlchemy!), I fixed it.

**Comment by mhauru** [comment]
I also silenced a warning coming from pandas that we were losing nanosecond precision when converting to python datetimes. With that we are down to 4 warnings, which all seem to come from statsmodels, and would probably require a deeper dive.

File: pyproject.toml (hunk 2 above)
Comment (EdwinB12) [reply pyproject.toml:92]: Ahh damn, good spot.

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (EdwinB12) [reply .github/workflows/tests.yaml:7]: No, I'll make that change.

File: pyproject.toml (hunk 2 above)
Comment (EdwinB12) [reply pyproject.toml:92]: Out of interest, what does this ignore do?

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (mhauru) [reply .github/workflows/tests.yaml:7]: Oh right, I only now realised this only concerns pull requests. That makes sense.

**Review by mhauru — APPROVED** [review]

**Pull request merged**